changelog-maker.js

```diff
@@ -122,22 +110,92 @@ function onCommitList (err, list) {
     if (!quiet) {
       printCommits(list)
     }
+
+    resolve()
+  })}) // eslint-disable-line brace-style, block-spacing
+}
+
+// convert a Stream into a ListStream, then List, then Promisify that List
+function streamToPromList (stream) {
+  return new Promise((resolve, reject) => {
+    stream.pipe(list.obj((err, list) => {
+      if (err) { reject(err) }
+      resolve(list)
+    }))
   })
 }
 
-let _startrefcmd = replace(refcmd, { ref: argv['start-ref'] || defaultRef })
-let _endrefcmd = argv['end-ref'] && replace(refcmd, { ref: argv['end-ref'] })
-let _sincecmd = replace(commitdatecmd, { refcmd: _startrefcmd })
-let _untilcmd = argv['end-ref'] ? replace(commitdatecmd, { refcmd: _endrefcmd }) : untilcmd
-let _gitcmd = replace(gitcmd, { sincecmd: _sincecmd, untilcmd: _untilcmd })
-
-debug('%s', _startrefcmd)
-debug('%s', _endrefcmd)
-debug('%s', _sincecmd)
-debug('%s', _untilcmd)
-debug('%s', _gitcmd)
-
-gitexec.exec(process.cwd(), _gitcmd)
-  .pipe(split2())
-  .pipe(commitStream(ghId.user, ghId.repo))
-  .pipe(list.obj(onCommitList))
+// get the start ref and end ref
+function getRefs () {
+  const prevTagCmd = 'git describe --abbrev=0 --tags $(git rev-list --tags --skip=1 --max-count=1)'
+  const latestTagCmd = 'git describe --abbrev=0 --tags'
+  const refCmd = 'git rev-list --max-count=1 {{ref}}'
+
+  const argSRef = argv['start-ref']
+  const argERef = argv['end-ref']
+  let startRefCmd = argSRef ? replace(refCmd, { ref: argSRef }) : prevTagCmd
+  let endRefCmd = argERef ? replace(refCmd, { ref: argERef }) : latestTagCmd
+  if (argv.a || argv.all) {
+    startRefCmd = 'git rev-list --max-parents=0 HEAD'
+    endRefCmd = 'git rev-list --max-count=1 HEAD'
+  }
+
+  return Promise.all([
+    streamToPromList(gitexec.exec(process.cwd(), startRefCmd)
+      .pipe(split2()))
+      .then((list) => list.join('\n')),
+    streamToPromList(gitexec.exec(process.cwd(), endRefCmd)
+      .pipe(split2()))
+      .then((list) => list.join('\n'))
+  ])
+}
+
+// get the ref's commit message
+function printRefMessage (ref) {
+  const messageCmd = `git log --format=%B -n 1 ${ref}`
+  return streamToPromList(gitexec.exec(process.cwd(), messageCmd)
+    .pipe(split2()))
+    .then((list) => {
+      if (quiet) { return }
+      // print the start ref's commit message in a code block if not simple
+      if (!simple) { console.log('```') }
+      printCommits(list)
+      if (!simple) { console.log('```') }
+    })
+}
+
+// print the changelog from start ref to end ref
+function printChangelog (startRef, endRef) {
+  const logCmd = `git log --pretty=full ${startRef}...${endRef}`
+  return streamToPromList(gitexec.exec(process.cwd(), logCmd)
+    .pipe(split2())
```

Comment:
might want to abstract out a function to reduce duplication of the `process.cwd()` and `split2()`